modules/packages/LinearAlgebra.chpl

```diff
@@ -937,6 +937,132 @@ proc trace(A: [?D] ?eltType) {
   return trace;
 }
 
+/* Type to choose which triangle of a matrix is desired */
+enum lowerOrUpper {
+  lower,
+  upper
+}
+
+
+/* Perform a Cholesky factorization on matrix ``A``.  ``A`` must be square.
+   Argument ``uplo`` indicates whether to return the lower or upper triangular
+   factor.  Matrix ``A`` is not modified.
+ */
+proc cholesky(A: [] ?t, uplo = lowerOrUpper.lower) where A.rank == 2 &&
+                                                         (isRealType(t) ||
+                                                          isComplexType(t)) {
+  assert(A.domain.dim(1) == A.domain.dim(2));
+  var copy = A;
+  const uploStr = if uplo == lowerOrUpper.lower then "L" else "U";
+  potrf(lapack_memory_order.row_major, uploStr, copy);
+
+  // tril and triu make/return an extra copy.  Should we zero the unused
+  // triangle of the array manually instead to avoid the copy?
+  return if uplo == lowerOrUpper.lower then tril(copy) else triu(copy);
+}
+
+
+/* Type to choose which eigenvectors to compute in the ``eigvals`` function. */
+enum EigenVecType {
```

Comment:
I feel similarly about this `enum` argument, but am not sure how it could map down to something simpler. Do you know how other libraries (such as scipy, matlab, julia, etc.) handle this argument?

Reply:
Julia and SciPy both have separate functions for computing only eigenvalues vs. both eigenvalues and eigenvectors.  Neither appears to have a direct way to get the left eigenvectors.

Matlab bases it on what the return value is assigned to:
```matlab
e = eig(A)       % just eigenvalues
[V,D] = eig(A)   % eigenvalues and right eigenvectors
[V,D,W] = eig(A) % eigenvalues, right eigenvectors, and left eigenvectors
```

Reply:
Thanks for looking into.

Matlab is pretty slick. I don't know of a way to do that in Chapel..

Splitting into separate functions like SciPy and Julia seems appealing to me. Do you think we could do something similar with our interface?

Reply:
I agree we can't do what Matlab does without a pretty significant language change, so that's out.

If you squint just right this is kind of already multiple functions since `rl` is a param.

If we were to break them up it would need to be four different functions with four different names.  I initially did have them split and overloaded, and which one got picked was based on the arguments, but as I simplified away arguments wasn't able to do that anymore, so I combined them to keep the overload feel.